tfhe/src/core_crypto/algorithms/lwe_multi_bit_programmable_bootstrapping.rs

```diff
@@ -0,0 +1,502 @@
+    }
+
+    assert!(work_queue.len() == lwe_mask.lwe_dimension().0 / bunching_factor.0);
+
+    let work_queue = Arc::new(Mutex::new(work_queue));
+
+    // At any point in time the idea is to have half the buffer already processed ready for the
+    // consumer and the other half being filled by the producer threads.
+    let ring_buffer_size: usize = thread_count.0 * 2;
+
+    let lut_poly_size = lut.polynomial_size();
+    let monomial_degree = pbs_modulus_switch(
+        *lwe_body.0,
+        lut_poly_size,
+        ModulusSwitchOffset(0),
+        LutCountLog(0),
+    );
+
+    // Modulus switching
+    lut.as_mut_polynomial_list()
+        .iter_mut()
+        .for_each(|mut poly| {
+            polynomial_wrapping_monic_monomial_div_assign(
+                &mut poly,
+                MonomialDegree(monomial_degree),
+            )
+        });
+
+    // TODO: (?) add a "next task idx" variable behind mutex + condvar to force the threads to
+    // generate the fourier GGSWs in the serial order (for determinism purposes), if a thread
+    // for whatever reason has a task index that does not match the next one that needs to
+    // execute then it will wait for the thread with the index to complete it's work and update
+    // the expected next task idx
+    let fourier_multi_bit_ggsw_buffers = Arc::new(
+        (0..ring_buffer_size)
+            .map(|_| {
+                (
+                    Mutex::new(false),
+                    Condvar::new(),
+                    Mutex::new(FourierGgswCiphertext::new(
+                        multi_bit_bsk.glwe_size(),
+                        multi_bit_bsk.polynomial_size(),
+                        multi_bit_bsk.decomposition_base_log(),
+                        multi_bit_bsk.decomposition_level_count(),
+                    )),
+                )
+            })
+            .collect::<Vec<_>>(),
+    );
+
+    thread::scope(|s| {
+        let produce_multi_bit_fourier_ggsw = || {
+            let mut buffers = ComputationBuffers::new();
+
+            let fft = Fft::new(multi_bit_bsk.polynomial_size());
+            let fft = fft.as_view();
+
+            buffers.resize(fft.forward_scratch().unwrap().unaligned_bytes_required());
+
+            let mut unit_polynomial =
+                Polynomial::new(Scalar::ZERO, multi_bit_bsk.polynomial_size());
+            unit_polynomial.as_mut()[0] = Scalar::ONE;
+            let mut a_monomial = unit_polynomial.clone();
+            let mut fourier_a_monomial = FourierPolynomial::new(multi_bit_bsk.polynomial_size());
+
+            let fourier_multi_bit_ggsw_buffers = fourier_multi_bit_ggsw_buffers.clone();
+
+            let work_queue = work_queue.clone();
+
+            loop {
+                let mut queue_lock = work_queue.lock().unwrap();
+                let maybe_work = queue_lock.pop();
+                drop(queue_lock);
+
+                if let Some((task_idx, lwe_mask_elements, ggsw_bunch)) = maybe_work {
```

Comment:
Since we are already requiring  rust 1.65 we could make use of let-else

```rust
let Some((task_idx, lwe_mask_elements, ggsw_bunch)) = maybe_work else {
 break;
}
```

Reply:
neat, I'll update